src/analyzer.fs

```diff
@@ -3,9 +3,155 @@
 open System.Collections.Generic
 open Ast
 
-// The module performs some static analysis on the code and stores the
-// information in the AST nodes, e.g. find which variables are modified,
-// which declarations can be inlined.
+// We can visit Var uses in evaluation order (sometimes twice: read then write),
+// and we know if they're read and/or written (by assignment operators or by in/out),
+// if it's a field access, and if it's a declaration initialization.
+
+type VarUse = {
+    access: Access
+    isFieldAccess: bool
+    isDecl: bool
+} with
+    override this.ToString() = $"{this.access}" + (if this.isFieldAccess then " field" else "") + (if this.isDecl then " decl" else " use")
+
+[<NoComparison; NoEquality>]
+type VarVisitor(onVarUse: VarUse -> Ident -> VarDecl -> unit) =
+
+    member this.onVisitVar = function
+        | ResolvedVariableUse (v, vd) -> onVarUse this.varUse v vd
+        | _ -> ()
+
+    member val varUse: VarUse = { access = { isWrite = false; isRead = true }; isFieldAccess = false; isDecl = false } with get, set
+
+    member this.using newContext go =
+        let old = this.varUse
+        this.varUse <- newContext
+        try
+            go()
+        finally
+            this.varUse <- old
+
+    // Vars in the AST can be a few things:
+    // [A] A declaration without initialization is neither a read nor a write.            int x;
+    // [B] A declaration with initialization is a "isDecl" write.                         int x = 1;
+    // [C] An assignment is a write.                                                      x = 1;
+    // [D] An augmented assignment is a read and then a write.                            x += 1;
+    // [E] A function name itself is neither a read nor a write.                          x();
+    // [F] A function call with an in parameter is a read.                                sin(x);
+    // [G] A function call with an out parameter is a write.                              void f(out p) { p = 1; } ... f(x);
```

Comment:
[E] Why is it not a read?
[F] rephrase: "A in parameter in a function call is a read" (the function call itself is not a var)
[G] [H] ditto